commit 10744b7b97ac0de054eb7573b97e28f7ce2544b8
Author: Palmer Dabbelt <palmer@dabbelt.com>
Date:   Sat May 18 20:58:22 2013 -0700

    Saves clipboard as an image
    
    This patch changes mupdf such that when the user presses the 'P' key
    it will save the current clipboard box as an image to "mupdf.pnm" in
    the current working directory.

diff --git a/apps/pdfapp.c b/apps/pdfapp.c
index 78b00ed..3697af4 100644
--- a/apps/pdfapp.c
+++ b/apps/pdfapp.c
@@ -80,6 +80,7 @@ char *pdfapp_usage(pdfapp_t *app)
 		"N\t\t-- find previous search result\n"
 		"c\t\t-- toggle between color and grayscale\n"
 		"i\t\t-- toggle inverted color mode\n"
+		"I\t\t-- dump image to PNM\n"
 		"q\t\t-- quit\n"
 	;
 }
@@ -1291,6 +1292,71 @@ void pdfapp_onkey(pdfapp_t *app, int c)
 		loadpage = 0;
 		break;
 
+	/*
+	 * Dumping images
+	 */
+
+	case 'I':
+	{
+		int x0, y0, x1, y1;
+		unsigned char *image_samples;
+		int image_w;
+#if 0
+                int image_h, image_n;
+#endif
+		FILE *file;
+		int i, j;
+
+		x0 = app->selr.x0;
+		y0 = app->selr.y0;
+		x1 = app->selr.x1;
+		y1 = app->selr.y1;
+
+		if (x0 < 0)
+			x0 = 0;
+		if (y0 < 0)
+			y0 = 0;
+		if (x1 >= app->winw)
+			x1 = app->winw-1;
+		if (y1 >= app->winh)
+			y1 = app->winh-1;
+
+		image_samples = fz_pixmap_samples(app->ctx, app->image);
+		image_w = fz_pixmap_width(app->ctx, app->image);
+#if 0
+		image_h = fz_pixmap_height(app->ctx, app->image);
+		image_n = fz_pixmap_components(app->ctx, app->image);
+#endif
+
+		file = fopen("mupdf.pnm", "w");
+		if (file == NULL)
+		{
+			fprintf(stderr, "Could not open 'mupdf.pnm'\n");
+			break;
+		}
+
+		fprintf(file, "P3\n%d %d\n%d\n", x1-x0+1, y1-y0+1, 255);
+
+		for (i = y0; i <= y1; i++)
+		{
+			for (j = x0; j <= x1; j++)
+			{
+				int index;
+
+				index = (i * image_w) + j;
+
+				fprintf(file, "%d %d %d\n",
+					image_samples[(4*index)+0],
+					image_samples[(4*index)+1],
+					image_samples[(4*index)+2]);
+			}
+		}
+
+		fclose(file);
+
+		break;
+	}
+
 	}
 
 	if (c < '0' || c > '9')
